File: backend/src/main/java/com/each/eirv/cinema_each/service/ClientesService.java
package com.each.eirv.cinema_each.service;

import com.each.eirv.cinema_each.dto.*;
import com.each.eirv.cinema_each.repository.ClientesRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ClientesService {

    private final ClientesRepository clientesRepository;

    // RF06 - Ranking de Clientes por Compra
    public List<ClientesRankingDTO> getRankingClientesPorCompra(LocalDate inicio, LocalDate fim) {
        return clientesRepository.rankingClientesPorCompra(inicio, fim);
    }

    // RF29 - Ranking por resgate
    public List<ClientesResgateRankingDTO> getRankingClientesPorResgate() {
        return clientesRepository.rankingClientesPorResgate();
    }

    // RF17 - Idade média por filme
    public List<IdadeMediaDTO> getIdadeMediaPorFilme() {
        return clientesRepository.consultarIdadeMediaPorFilme();
    }

    // RF07 - Idade média por gênero
    public List<IdadeMediaDTO> getIdadeMediaPorGenero() {
        return clientesRepository.consultarIdadeMediaPorGenero();
    }

    // RF11 - Público estudantil
    public List<PublicoEstudantilDTO> getPublicoEstudantilRanking(LocalDate inicio, LocalDate fim) {
        return clientesRepository.consultarPublicoEstudantil(inicio, fim);
    }

    // RF14 - Clientes sem meia
    public List<ClientesSimplesDTO> getClientesSemMeiaEntrada() {
        return clientesRepository.consultarClientesSemMeiaEntrada();
    }

    // RF10 - Clientes por Filme
    public List<ClientesSimplesDTO> getClientesPorFilme(Long idFilme) {
        if (idFilme == null || idFilme <= 0) {
            throw new IllegalArgumentException("ID do Filme inválido.");
        }

        return clientesRepository.consultarClientesPorFilme(idFilme);
    }

    // RF19 - Gêneros mais assistidos por mês
    public List<GeneroMensalDTO> getGenerosMaisAssistidosPorMes(LocalDate inicio, LocalDate fim) {
        return clientesRepository.consultarGenerosMaisAssistidosPorMes(inicio, fim);
    }
}

File: frontend/src/features/clientes/api/clientesService.ts
import { get } from "../../../api/api";
import {
  type ClienteRanking,
  type ClienteResgateRanking,
  type IdadeMedia,
  type PublicoEstudantil,
  type ClienteSimples,
  type GeneroMensal,
} from "../../../types/clientesTypes";


//RF06 - Ranking de Clientes por Compra
export async function getRankingClientesPorCompra(params?: {
  inicio?: string;
  fim?: string;
}): Promise<ClienteRanking[]> {
  const query = params
    ? `?inicio=${params.inicio ?? ""}&fim=${params.fim ?? ""}`
    : "";
  return await get(`/api/publico/clientes/ranking/compras${query}`);
}

//RF29 - Ranking por resgates de benefícios
export async function getRankingClientesPorResgate(): Promise<ClienteResgateRanking[]> {
  return await get(`/api/publico/clientes/ranking/resgates`);
}

//RF17 - Idade média dos clientes por filme
export async function getIdadeMediaPorFilme(): Promise<IdadeMedia[]> {
  return await get(`/api/publico/idade-media/filmes`);
}

//RF07 - Idade média dos clientes por gênero de filme
export async function getIdadeMediaPorGenero(): Promise<IdadeMedia[]> {
  return await get(`/api/publico/idade-media/generos`);
}

//RF11 - público estudantil
export async function getPublicoEstudantil(params?: {
  inicio?: string;
  fim?: string;
}): Promise<PublicoEstudantil[]> {
  const query = params
    ? `?inicio=${params.inicio ?? ""}&fim=${params.fim ?? ""}`
    : "";
  return await get(`/api/publico/estudantil/ranking${query}`);
}

//RF14 - Clientes sem meia-entrada
export async function getClientesSemMeia(): Promise<ClienteSimples[]> {
  return await get(`/api/publico/clientes/sem-meia`);
}

//RF10 - Clientes por filme
export async function getClientesPorFilme(idFilme: number): Promise<ClienteSimples[]> {
  return await get(`/api/publico/clientes/por-filme?idFilme=${idFilme}`);
}

//RF19 - Gêneros de filme mais assistidos mensalmente
export async function getGenerosMaisAssistidos(params?: {
  inicio?: string;
  fim?: string;
}): Promise<GeneroMensal[]> {
  const query = params
    ? `?inicio=${params.inicio ?? ""}&fim=${params.fim ?? ""}`
    : "";
  return await get(`/api/publico/generos/mensal${query}`);
}